File: tools/generate-category-pages.js
#!/usr/bin/env node
// One-time generator that stamps category page shells from a template.
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const CATALOG = path.join(ROOT, 'js', 'catalog.js');
const CONFIG = path.join(ROOT, 'js', 'config.js');

// Make catalog.js attach to a global window object
const sandbox = { window: {} };
const vm = require('vm');
const ctx = vm.createContext(sandbox);
vm.runInContext(fs.readFileSync(CONFIG, 'utf8'), ctx);
vm.runInContext(fs.readFileSync(CATALOG, 'utf8'), ctx);
const { CATEGORY_META, BASE_DESIGNS, SITE_CONFIG } = sandbox.window;

const TEMPLATE = fs.readFileSync(path.join(__dirname, 'category-template.html'), 'utf8');

if (!CATEGORY_META || !BASE_DESIGNS || !SITE_CONFIG) {
  console.error('Could not load category or site metadata');
  process.exit(1);
}

function escapeHtml(value) {
  return String(value).replaceAll('&', '&amp;').replaceAll('"', '&quot;').replaceAll('<', '&lt;').replaceAll('>', '&gt;');
}

function categoryDesigns(key, info) {
  const keys = info.aggregate ? info.subcategories : [key];
  return keys.flatMap((category) => (BASE_DESIGNS[category] || []).map((design) => design.name));
}

function structuredData(key, info, canonical, seoImage) {
  const breadcrumbItems = [{ name: 'Home', item: `${SITE_CONFIG.siteUrl}/` }];
  if (info.parent) {
    breadcrumbItems.push({
      name: CATEGORY_META[info.parent].title,
      item: `${SITE_CONFIG.siteUrl}/${info.parent}.html`
    });
  }
  breadcrumbItems.push({ name: info.title, item: canonical });
  const designs = categoryDesigns(key, info);
  return JSON.stringify({
    '@context': 'https://schema.org',
    '@graph': [
      {
        '@type': 'CollectionPage',
        '@id': `${canonical}#webpage`,
        url: canonical,
        name: info.seoTitle,
        description: info.seoDescription,
        isPartOf: { '@id': `${SITE_CONFIG.siteUrl}/#website` },
        primaryImageOfPage: { '@type': 'ImageObject', url: seoImage },
        breadcrumb: { '@id': `${canonical}#breadcrumb` },
        mainEntity: { '@id': `${canonical}#catalog` }
      },
      {
        '@type': 'BreadcrumbList',
        '@id': `${canonical}#breadcrumb`,
        itemListElement: breadcrumbItems.map((item, index) => ({
          '@type': 'ListItem',
          position: index + 1,
          name: item.name,
          item: item.item
        }))
      },
      {
        '@type': 'ItemList',
        '@id': `${canonical}#catalog`,
        name: `${info.title} designs`,
        numberOfItems: designs.length,
        itemListElement: designs.map((name, index) => ({
          '@type': 'ListItem',
          position: index + 1,
          name: name
        }))
      }
    ]
  }).replaceAll('<', '\\u003c');
}

const generated = [];
Object.keys(CATEGORY_META).forEach((key) => {
  const info = CATEGORY_META[key];
  const canonical = `${SITE_CONFIG.siteUrl}/${key}.html`;
  const seoImage = `${SITE_CONFIG.siteUrl}/${info.seoImage}`;
  const replacements = {
    key,
    title: info.title,
    description: info.description,
    image: 'images/nyravi_logo.png',
    seoTitle: info.seoTitle,
    seoDescription: info.seoDescription,
    canonical,
    seoImage,
    structuredData: structuredData(key, info, canonical, seoImage)
  };
  const page = Object.entries(replacements).reduce(
    (output, [name, value]) => output.replaceAll(`{${name}}`, name === 'structuredData' ? value : escapeHtml(value)),
    TEMPLATE
  );
  fs.writeFileSync(path.join(ROOT, `${key}.html`), page);
  generated.push(`${key}.html`);
});

console.log('Generated:', generated.join(', '));

File: tools/generate-category-pages.py
#!/usr/bin/env python3
"""One-time generator that stamps category page shells from a template."""
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
CATALOG = ROOT / "js" / "catalog.js"
TEMPLATE = """<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>{title} | Nyravi Boutique</title>
  <meta name="description" content="{description}">
  <link rel="icon" href="images/favicon.ico">
  <link rel="stylesheet" href="css/main.css">
</head>
<body data-page="{key}">
  <header id="site-header"></header>
  <main>
    <section class="page-hero">
      <div class="page-hero-bg" style="background-image:url('{image}')"></div>
      <div class="page-hero-content">
        <h1 id="category-title">{title}</h1>
        <p id="category-description">{description}</p>
      </div>
    </section>
    <section class="category-main">
      <div class="container">
        <div id="subcategory-grid" class="subcategory-grid"></div>
        <div class="category-intro">
          <h2>Choose your style</h2>
          <p>Browse our designs, pick your tier and order directly on WhatsApp.</p>
        </div>
        <div id="tier-tabs" class="tier-tabs"></div>
        <div id="product-grid" class="product-grid"></div>
      </div>
    </section>
  </main>
  <footer id="site-footer"></footer>
  <script src="js/config.js"></script>
  <script src="js/catalog.js"></script>
  <script src="js/components.js"></script>
  <script src="js/app.js"></script>
</body>
</html>
"""

def parse_meta_js(path):
    """Very small parser: extracts CATEGORY_META object from the catalog file."""
    text = path.read_text()
    start = text.find("const CATEGORY_META = {")
    if start == -1:
        raise ValueError("CATEGORY_META not found")
    brace = start + len("const CATEGORY_META = ")
    end = text.find("};", brace)
    obj_text = text[brace:end + 1]
    # Replace placeholderImage(...) calls with a dummy URL string so JSON can parse
    import re
    obj_text = re.sub(r'placeholderImage\([^)]*\)', '""', obj_text)
    # Remove trailing commas that may precede closing braces/brackets
    obj_text = re.sub(r',(\s*[}\]])', r'\1', obj_text)
    return json.loads(obj_text)

if __name__ == "__main__":
    meta = parse_meta_js(CATALOG)
    generated = []
    for key, info in meta.items():
        image = "images/nyravi_logo.png"  # will be replaced by JS; keep fallback
        page = TEMPLATE.format(
            key=key,
            title=info.get("title", key),
            description=info.get("description", ""),
            image=image,
        )
        out = ROOT / f"{key}.html"
        out.write_text(page)
        generated.append(out.name)
    print("Generated:", generated)
